demos/unsupported/cstext/texteditor/texteditor.man
.\" $XConsortium: texteditor.man /main/4 1995/07/17 10:48:58 drk $
.\" Motif
.\"
.\" Copyright (c) 1987-2012, The Open Group. All rights reserved.
.\"
.\" These libraries and programs are free software; you can
.\" redistribute them and/or modify them under the terms of the GNU
.\" Lesser General Public License as published by the Free Software
.\" Foundation; either version 2.1 of the License, or (at your option)
.\" any later version.
.\"
.\" These libraries and programs are distributed in the hope that
.\" they will be useful, but WITHOUT ANY WARRANTY; without even the
.\" implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
.\" PURPOSE. See the GNU Lesser General Public License for more
.\" details.
.\"
.\" You should have received a copy of the GNU Lesser General Public
.\" License along with these librararies and programs; if not, write
.\" to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
.\" Floor, Boston, MA 02110-1301 USA
...\" 
...\" 
...\" HISTORY
...\" "
.TH texteditor 1X MOTIF "Demonstration programs"
.SH NAME
\*Ltexteditor\*O\(emedit text with multiple views and renditions.
.SH SYNOPSIS
.sS
\*Ltexteditor\*O
.sE
.SH DESCRIPTION
\*Ltexteditor\*O 
is a primitive text editor, based on Motif's XmCSText
widget.  In addition to standard text editing features, it supports
multiple views of the same underlying text, modification of the rendition
attributes of individual pieces of text, and the reading and writing of
multiple file formats.
.SH USAGE
This section describes how to use the menus and dialog boxes.
.SH "File Menu"
This is a completely standard File menu.  The only important note is that
the Print... command only supports text-only printing.
.SH "Edit Menu"
This is a completely standard Edit menu.
.SH "View Menu"
The View menu controls the views of the text.  It includes:
.IP "\fBSplit Pane\fP"
to add a new independently scrollable view of the text
.IP "\fBRemove Pane\fP"
to remove a designated pane
.IP "\fBOne Pane\fP"
to remove all panes except the designated pane
.PP
The designated pane is determined in the following way.  Using an explicit
focus policy, it is the last view which had focus.  Using a pointer focus
policy, it is the last view which the user interacted with.
.SH "Options Menu"
.IP "\fBKeep File Dialog Open\fP"
If this option is selected, the file selection dialog will remain open
after an operation is completed.
.IP "\fBRevert Dialog to Open\fP"
If this option is selected, the file selection dialog's operation will
automatically be reset to \fBOpen\fP after an operation is completed.
.SH "Renditions Menu"
This menu provides commands to change the appearance of selected text.  
.SH "Help Menu"
This is a completely standard Help menu.
.SH "FileSelectionBox"
The FileSelectionBox that is displayed when the \fBOpen...\fP, \fBSave\fP,
or \fBSave As...\fP commands are chosen is standard with the following additions.
The \fBFormat\fP menu allows text to be read or written in the following
formats:
.IP "\fBText Only\fP"
Reads in the file as straight ASCII text.  Writes only the text portion of
a file.  All rendition information is lost.
.IP "\fBByte Stream\fP"
Reads and writes files in Compound String Byte Stream format.  This
preserves all rendition information.  See 
.IR XmCvtXmStringToByteStream(3X)
and
.IR XmCvtByteStreamToXmString(3X).
.IP "\fBXmdHelp Markup\fP"
Reads and writes files in using the XmdHelp Markup language.  This is a
simple quasi-SGML used by the Xmd Help widget.  See the onHelp application
in demos/lib/Xmd for more information.
.PP
The radio box immediately below the \fBFormat\fP menu allows interactively
switching the operation to be performed between \fBOpen\fP, \fBSave\fP,
\fBCopy\fP, and \fBMove\fP.
.SH "X Defaults"
The class of the application is \fIXmdTextEditor\fP.  The widget of primary
interest is the XmCSText widget.
.PP
In particular, users familiar with emacs may want to add the following translations
to their .Xdefaults files:

.nf
.ta .25i 2i
*XmCSText.translations: #override\\n\\
    Mod1<Btn3Down>:	scroll-cursor-vertically()\\n\\
  Mod1<Btn3Motion>:	scroll-cursor-vertically()\\n\\
	Ctrl<Key>a:	beginning-of-line()\\n\\
	Ctrl<Key>b:	backward-character()\\n\\
	Ctrl<Key>d:	delete-next-character()\\n\\
	Ctrl<Key>e:	end-of-line()\\n\\
	Ctrl<Key>f:	forward-character()\\n\\
	Ctrl<Key>g:	beep()\\n\\
	Ctrl<Key>h:	delete-previous-character()\\n\\
	Ctrl<Key>i:	cut-primary()\\n\\
	Ctrl<Key>j:	newline-and-indent()\\n\\
	Ctrl<Key>k:	set-anchor() end-of-line() key-select() cut-clipboard()\\n\\
	Ctrl<Key>l:	redraw-display()\\n\\
	Ctrl<Key>m:	newline()\\n\\
	Ctrl<Key>n:	next-line()\\n\\
	Ctrl<Key>o:	newline-and-backup()\\n\\
	Ctrl<Key>p:	previous-line()\\n\\
	Ctrl<Key>v:	next-page()\\n\\
	Ctrl<Key>w:	cut-clipboard()\\n\\
	Ctrl<Key>y:	paste-clipboard()\\n\\
	Ctrl<Key>z:	scroll-one-line-up()\\n\\
	Mod1<Key>b:	backward-word()\\n\\
	Mod1<Key>d:	delete-next-word()\\n\\
	Mod1<Key>f:	forward-word()\\n\\
	Mod1<Key>h:	delete-previous-word()\\n\\
	Mod1<Key>i:	copy-primary()\\n\\
	Mod1<Key>k:	delete-to-end-of-line()\\n\\
	Mod1<Key>v:	previous-page()\\n\\
	Mod1<Key>w:	copy-clipboard()\\n\\
	Mod1<Key>z:	scroll-one-line-down()\\n\\
	Mod1 Shift<Key>greater:	end-of-file()\\n\\
	Mod1<Key>less:	beginning-of-file()\\n\\
	Mod1<Key>]:	forward-paragraph()\\n\\
	Mod1<Key>[:	backward-paragraph()
.fi

.SH FILES
.IP ./demo1.txt
A Text Only format file.
.IP ./demo1.bts
A Byte Stream format file.  Contains the text in demo1.txt with additional
renditions.
.IP ./demo1.help
An XmdHelp Markup format file.  Produces the same output as demo1.bts.

.SH BUGS
Moving and removing files are not currently implemented.  The following
menu items are not implemented:  \fBPrint\fP, \fBPrint Setup...\fP, \fBClose\fP, \fBUndo\fP, \fBRepeat\fP,
\fBSelect All\fP, \fBDeselect All\fP, and all the \fBHelp\fP menu items.  Changing the
rendition of text contained within another rendition change can sometimes
cause renditions to "bleed" into the following text.  Changing the rendition
of a line selected by a triple-click will cause all lines until the next
rendition change to also get the selected rendition.  WYSIWYG editing and
saving of XmdHelp Markup format is not supported.  The selection of
renditions is limited.


.attic/ChangeLog
2017-11-10 11:49:10 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1665 (HEAD, origin/master, origin/HEAD, master)

2017-10-30 23:04:15 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1651

2017-10-30 22:22:31 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1650

2017-10-30 21:47:21 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1652

2017-08-31 06:28:27 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1655

2017-08-28 05:33:53 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1625

2017-08-26 06:12:14 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1656

2017-08-23 04:10:55 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1658

2017-08-22 06:05:00 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1659

2017-08-22 03:27:36 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1660

2017-03-28 01:40:33 +0300 Oleksiy Chernyavskyy 

	* updated version and release data

2017-03-28 01:32:30 +0300 Oleksiy Chernyavskyy 

	* removed redundant includes

2017-03-28 01:11:59 +0300 Oleksiy Chernyavskyy 

	* removed bugfix 1565. See details on http://bugs.motifzone.net/show_bug.cgi?id=1565

2016-11-09 07:19:53 +0200 Alex 

	* obsolete malloc.h changed on stdlib.h

2016-06-11 05:03:21 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-11 05:02:55 +0300 Oleksiy Chernyavskyy 

	* updated release 2.3.6 information

2016-06-11 04:57:54 +0300 Oleksiy Chernyavskyy 

	* fixid parallel building (make -j) error

2016-06-11 04:57:20 +0300 Oleksiy Chernyavskyy 

	* removed automatic configure start from autogen.sh

2016-06-11 04:26:48 +0300 Oleksiy Chernyavskyy 

	* minor fix

2016-06-11 04:21:53 +0300 Oleksiy Chernyavskyy 

	* removed extra dist Imakefile from all Makefile.am files

2016-06-11 04:06:37 +0300 Oleksiy Chernyavskyy 

	* Purified copyright headers (Motif is LGPL covered). Removed many Imakefile's and Imake-related stuff

2016-06-10 23:40:43 +0300 Oleksiy Chernyavskyy 

	* updated release notes and version 2.3.5 -> 2.3.6

2016-06-10 23:27:56 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-10 23:15:00 +0300 Oleksiy Chernyavskyy 

	* bug 1654 fix

2016-06-10 21:58:16 +0300 Oleksiy Chernyavskyy 

	* old patches. need review

2016-03-29 03:48:59 +0300 Oleksiy Chernyavskyy 

	* removed obsolete files

2016-03-17 04:23:17 +0200 Oleksiy Chernyavskyy 

	* release 2.3.5 (tag: release_2_3_5)

2016-03-16 23:07:12 +0200 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-03-16 23:06:24 +0200 Oleksiy Chernyavskyy 

	* added mkchangelog script for automated generation of the ChangeLog

2016-03-16 00:46:49 +0200 Oleksiy Chernyavskyy 

	* Reimplemented bugfix 1565

2016-02-26 03:36:51 +0200 Oleksiy Chernyavskyy 

	* Updated RPM creation instructions

2014-08-20 15:45:41 +0300 Oleksiy Chernyavskyy 

	* fixed bug 1636

2013-08-21 11:49:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1612 (Label size computed wrong within a Form).

2013-07-11 18:14:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1618 (Eliminates 29 warnings of some type in modules from "demos" catalog).

2013-07-11 15:20:54 +0300 Mykola Vshyvkov 

	* Fixed bug #1616 (Eliminates 23 warnings "suggest parentheses around assignment used" in modules from "demos" catalog).

2013-07-11 15:01:50 +0300 Mykola Vshyvkov 

	* Fixed bug #1617 (Eliminates 23 warnings "cast to pointer from integer of different size" in modules from "demos" catalog).

2013-07-11 12:35:30 +0300 Mykola Vshyvkov 

	* Fixed bug #1615 (Eliminates 35 warnings "enumeration value XXX not handled" in modules from "demos" catalog).

2013-07-10 06:55:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1614 (Eliminates 36 warnings "variable may be used uninitialized" in modules from "demos" catalog).

2013-07-10 05:05:28 +0300 Mykola Vshyvkov 

	* Fixed bug #1613 (Eliminates 60 warnings "implicit declaration of function" in modules from "demos" catalog).

2013-07-04 14:08:09 +0300 Mykola Vshyvkov 

	* Fixed bug #1611 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls in modules of "clients" catalog).

2013-07-03 21:27:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1610 (Eliminates 15 warnings of some types in modules of "clients" catalog).

2013-07-03 18:52:21 +0300 Mykola Vshyvkov 

	* Fixed bug #1609 (Eliminates 18 warnings  of "incompatible implicit declaration" in modules of "clients" catalog).

2013-07-03 14:58:22 +0300 Mykola Vshyvkov 

	* Fixed bug #1608 (Eliminates 30 warnings  of "array subscript has type 'char'" in modules of "clients" catalog).

2013-07-03 14:35:04 +0300 Mykola Vshyvkov 

	* Fixed bug #1607 (Eliminates 92 warnings  of "variable may be used uninitialized" in modules of "clients" catalog).

2013-07-03 12:48:55 +0300 Mykola Vshyvkov 

	* Fixed bug #1606 (Eliminates 130 warnings  of "expects argument of type" in modules of "clients" catalog).

2013-07-02 13:40:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 20:52:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 19:07:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1604 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls).

2013-07-01 19:02:25 +0300 Mykola Vshyvkov 

	* Fixed bug #1605 (Eliminates 14 warnings of some types in WML tool).

2013-06-26 23:45:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1603 (Eliminates 18 warnings of some types in some modules).

2013-06-26 19:21:20 +0300 Mykola Vshyvkov 

	* Fixed bug #1602 (Eliminates 6 warnings "implicit declaration of function" in some modules).

2013-06-26 18:36:47 +0300 Mykola Vshyvkov 

	* Fixed bug #1601 (Eliminates 6 warnings: "suggest explicit braces to avoid ambiguous" in some modules).

2013-06-26 17:40:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1600 (Eliminates 26 warnings about using uninitialized variables in some modules).

2013-06-26 15:34:53 +0300 Mykola Vshyvkov 

	* Fixed bug #1599 (Eliminates 14 warnings of incompatible type argument from some modules).

2013-06-25 17:37:02 +0300 Mykola Vshyvkov 

	* Updated ChangeLog.

2013-06-25 13:57:31 +0300 Mykola Vshyvkov 

	* Merge branch 'openmotif_2_3' of ssh://git.code.sf.net/p/motif/code into openmotif_2_3


demos/unsupported/cstext/texteditor/texteditor.c
/* 
 * Motif
 *
 * Copyright (c) 1987-2012, The Open Group. All rights reserved.
 *
 * These libraries and programs are free software; you can
 * redistribute them and/or modify them under the terms of the GNU
 * Lesser General Public License as published by the Free Software
 * Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * These libraries and programs are distributed in the hope that
 * they will be useful, but WITHOUT ANY WARRANTY; without even the
 * implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE. See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with these librararies and programs; if not, write
 * to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
 * Floor, Boston, MA 02110-1301 USA
*/ 
/* 
 * HISTORY
*/ 
#ifdef REV_INFO
#ifndef lint
static char rcsid[] = "$XConsortium: texteditor.c /main/6 1995/07/14 09:44:58 drk $"
#endif
#endif

/************************************************************
 *
 *      textedit.c -- simple text editor
 *              Ellis S. Cohen
 *         Open Software Foundation
 *
 *   This simple text editor allows a single file at a time to
 *   be edited in one or more panes, and also allows the file to
 *   be saved, saved under a new name, copied, moved, or removed
 *
 *   The code in this file is used to initialize the application,
 *   to define localizable strings, and to create widgets.
 *   It does so by using the Mrm library to read UID files (compiled
 *   UIL files).  The code in this file hides Mrm from the rest
 *   of the application.
 *
 ************************************************************/

#include <X11/Intrinsic.h>

#include <Xm/Xm.h>
#include <Xm/CSText.h>
#include <Xm/FileSB.h>

#include <Mrm/MrmPublic.h>

#include <Xmd/Help.h>
#include <Xmd/Print.h>

#include "dlg.h"
#include "tk.h"

extern char *appString;
extern XmString openFileString, openString,
                saveFileString, saveString,
                copyFileString, copyString,
                moveFileString, moveString,
                warnOpenString, warnWriteString, warnRemoveString;

XtAppContext app_context;
extern Display *dpy;
extern Widget toplevel, panedWindow, textFirst, textStore;
extern Widget fileDialog, saveDialog, warnDialog, questionDialog;
extern Widget helpDialog, printDialog;
extern Widget openToggle, saveToggle, copyToggle, moveToggle;
extern Widget formatMenu;

static  char *mrmfiles[] = { "textedit.uid" };
static XmRenderTable renditions;

#define MAX_ARGS 10
static Arg args[MAX_ARGS];
static int n;

static MrmHierarchy muid;
static MrmType type;

#define APP_CLASS "XmdTextEditor"
char *fallback_resources[] = {
  "*text.rows: 24",
  "*text.columns: 80",
  "*print_manager.printerList: lp,./texteditor.txt",
  "*selectColor:	XmDEFAULT_SELECT_COLOR",
  NULL
  };


 /************************************************************
 *     Create Text Pane
 ************************************************************/

      /* Called to create another text pane in the PanedWindow.
         "paneht" is its desired height. */

Widget
DemoTextPaneCreate(int paneht)
{
  Widget w = NULL;

  /* Each time "textPane" is fetched, a new text widget is created.
     Its source is obtained from textStore.
     We set paneMinimum (as well as height) to paneHt to ensure
     it will actually get that size */

  n = 0;
  XtSetArg( args[n], XmNsource, XmCSTextGetSource( textStore ) ); n++; 
  XtSetArg( args[n], XmNheight, paneht ); n++;
  XtSetArg( args[n], XmNpaneMinimum, paneht ); n++;
  XtSetArg( args[n], XmNrenderTable, renditions ); n++;
  MrmFetchWidgetOverride( muid, "textPane", panedWindow,
			 NULL, args, n,  &w, &type);
  
  XtManageChild( w );

  /* Now that "textPane" is managed, we can reset paneMinimum */

  n = 0;
  XtSetArg( args[n], XmNpaneMinimum, 20 ); n++;
  XtSetValues( XtParent(w), args, n );
  return w;
}

/************************************************************
 *     Initialize the Application
 ************************************************************/

int
main(int argc, char *argv[])
{
    Widget mainWindow;
    Widget viewMenu;
    Widget tmp_wid;
    Widget menuBar;

      /* Initialize Mrm and Xt */

    MrmInitialize();

    toplevel = XtVaAppInitialize(&app_context, APP_CLASS, NULL, 0, &argc, argv,
				 fallback_resources, 0);
    dpy = XtDisplay(toplevel);

      /* Associate callback names to be obtained from UIL files
         with callback routines in the application, and open
         the UIL database */

    TkRegisterCallbacks();
    MrmOpenHierarchy( 1, mrmfiles, NULL, &muid );

      /* Get string names (which may be locale-specific) */

    MrmFetchLiteral( muid, "appString", dpy, (XtPointer)&appString, &type );
    MrmFetchLiteral( muid, "openFileString", dpy, (XtPointer)&openFileString, &type );
    MrmFetchLiteral( muid, "openLabelString", dpy, (XtPointer)&openString, &type );
    MrmFetchLiteral( muid, "saveFileString", dpy, (XtPointer)&saveFileString, &type );
    MrmFetchLiteral( muid, "saveLabelString", dpy, (XtPointer)&saveString, &type );
    MrmFetchLiteral( muid, "copyFileString", dpy, (XtPointer)&copyFileString, &type );
    MrmFetchLiteral( muid, "copyLabelString", dpy, (XtPointer)&copyString, &type );
    MrmFetchLiteral( muid, "moveFileString", dpy, (XtPointer)&moveFileString, &type );
    MrmFetchLiteral( muid, "moveLabelString", dpy, (XtPointer)&moveString, &type );
    MrmFetchLiteral( muid, "warnOpenString", dpy, (XtPointer)&warnOpenString, &type );
    MrmFetchLiteral( muid, "warnWriteString", dpy, (XtPointer)&warnWriteString, &type );
    MrmFetchLiteral( muid, "warnRemoveString", dpy, (XtPointer)&warnRemoveString, &type );

      /* Instantiate and Manage MainWindow widget hierarchy */

    mainWindow = NULL;
    /* Need to exit cleanly here if this fails. */
    MrmFetchWidget( muid, "mainWindow", toplevel, &mainWindow, &type);
    if (mainWindow == NULL) exit(1);
    
    /* Modify the menuBar from its initial state. */
    menuBar = XtNameToWidget(mainWindow, "*menuBar");
    viewMenu = XtNameToWidget(menuBar, "*viewMenu");
    
    XtUnmanageChild(XtNameToWidget(viewMenu, "*changeViewButton"));
    XtUnmanageChild(XtNameToWidget(viewMenu, "*newWindowButton"));
    
    MrmFetchWidget(muid, "splitButton", viewMenu, &tmp_wid, &type);
    XtManageChild(tmp_wid);
    MrmFetchWidget(muid, "removePaneButton", viewMenu, &tmp_wid, &type);
    XtManageChild(tmp_wid);
    MrmFetchWidget(muid, "onePaneButton", viewMenu, &tmp_wid, &type);
    XtManageChild(tmp_wid);

    XtManageChild( mainWindow );
    XtRealizeWidget( toplevel );

      /* Locate the PanedWindow and its initial text pane */

    panedWindow = XtNameToWidget( mainWindow, "*panedWindow" );
    textFirst = XtNameToWidget( panedWindow, "*textFirst" );

    /* Updated renderTable based on Renditions menu. */
    XtVaGetValues(textFirst, XmNrenderTable, &renditions, NULL);
    renditions = TkUpdateRenditions(textFirst, renditions);

    XtVaSetValues(textFirst, XmNrenderTable, renditions, NULL);
    
    XtVaSetValues(XtNameToWidget(mainWindow, "*colorsMenu"),
		  XmNrenderTable, renditions, NULL);
    
      /* The unmanaged text widget textStore will hold onto the
         "source" shared by all the text panes */

/* $$$ comment the following code out */

    n = 0;
    XtSetArg( args[n], XmNsource, XmCSTextGetSource( textFirst ) ); n++; 
    textStore = NULL;
    MrmFetchWidgetOverride( muid, "textStore", toplevel,
                                  NULL, args, n,  &textStore, &type);

    XtRealizeWidget( textStore );

/* $$$ and replace it by

    textStore = textFirst;

    $$$ to get errors without the complication of shared sources */

      /* Instantiate FileSelectionBox widget hierarchy, but leave unmanaged */

    fileDialog = NULL;
    MrmFetchWidget( muid, "fileDialog", toplevel, &fileDialog, &type );

      /* Locate various toggle buttons placed in the FileSelectionBox,
         so that their toggle state can be set when necessary */

    openToggle = XtNameToWidget( fileDialog, "*openToggle" );
    saveToggle = XtNameToWidget( fileDialog, "*saveToggle" );
    copyToggle = XtNameToWidget( fileDialog, "*copyToggle" );
    moveToggle = XtNameToWidget( fileDialog, "*moveToggle" );

    /* Locate format menu so that it's current state can be read. */
    formatMenu = XtNameToWidget(fileDialog, "*formatMenu");
    
      /* Instantiate other dialogs, but leave unmanaged */

    saveDialog = NULL;
    warnDialog = NULL;
    questionDialog = NULL;
    MrmFetchWidget( muid, "saveDialog", toplevel, &saveDialog, &type );
    MrmFetchWidget( muid, "warnDialog", toplevel, &warnDialog, &type );
    MrmFetchWidget( muid, "questionDialog", toplevel, &questionDialog, &type );

    /* Help Dialog */
    n = 0;
    XtSetArg(args[n], XmNrenderTable, renditions); n++;
    XtSetArg(args[n], XmdNhelpFile, "texteditor"); n++;
    helpDialog = (Widget)XmdCreateHelpDialog(toplevel, "HelpDialog", args, n);
    
    /* Print Dialog */
    printDialog = XmdCreatePrintDialog(toplevel, "PrintDialog", NULL, 0);
    XtAddCallback(printDialog, XmdNprintCallback,
		  (XtCallbackProc)DlgPrintCB, NULL);

      /* Initialize the toolkit-specific layer */

    TkInit();

      /* Wait for events to trigger callbacks */

    XtAppMainLoop(app_context);
}



